rfcs/20190829-tfx-container-component-execution.md

```diff
@@ -0,0 +1,555 @@
+    # Select launcher by platform config.
+    for platform_config in component_platform_configs:
+      for launcher in self.launchers:
+        if launcher.can_launch(component.executor_spec, platform_config):
+          return ComponentLaunchInfo(component, launcher, platform_config)
+    for launcher in self.launchers:
+      if launcher.can_launch(component.executor_spec):
+        return ComponentLaunchInfo(component, launcher)
+
+  def run(self, pipeline) -> Optional[Any]:
+    component_launcher_infos = {c.name: self._get_component_launch_info(c)
+                                for c in pipeline.components)}
+    return self._run(self, pipeline, component_launcher_infos)
+
+  @abc.abstractmethod
+  def _run(self, pipeline, component_launcher_infos) -> Optional[Any]:
+    pass
+```
+
+### Output interface
+
+User container can receive a
+[tmp directory path from default artifact store](https://github.com/tensorflow/community/blob/2c0b009ef955975b15a3cc18b1378e0ed38f394e/rfcs/20190904-tfx-generic-container-based-component.md#artifact-properties-after-execution-is-complete)
+to write output data. The directory parameter will be called
+`exec_properties.tmp_path`, which can be passed in as a command line argument.
+The executor will look for `output.json` file under `exec_properties.tmp_path`
+to get the outputs from the component. The output file follows the following
+schema:
+
+```yaml
+"$id": https://pipeline.mlx.org/output.schema.json"
+"$schema": http://json-schema.org/draft-07/schema#"
+type: object
+title: Output
+properties:
+  error_status: { "$ref": "#/definitions/OutputErrorStatus" }
+  outputs:
+    type: object
+  exec_properties:
+    type: object
+definitions:
+  OutputErrorStatus:
+    type: object
+    properties:
+      code:
+        type: string
+        enum: [PERMANENT_ERROR, RETRYABLE_ERROR]
+      message:
+        type: string
+```
+
+The output.json file is optional, but if user’s container writes to the file. It
+overrides the default handling of k8s pod launcher. Here are the explanation of
+the output fields:
+
+*   error_status: it tells the executor whether it should retry or fail.
+*   outputs and exec_properties: they will be used to override the execution and
+    output artifact metadata in MLMD.
+
+The output interfaces relies on `BaseComponentLauncher` to update states back to
+MLMD from executor.
+
+### Auth context resolution
+
+K8s pod launcher internally uses k8s python client. The auth context resolution
+logic is as follows:
+
+1.  If the current env is in a cluster, uses `load_incluster_config` to load k8s
+    context.
+1.  If not, use default k8s active context to connect to remote cluster.
+
+### Pod launcher resiliency
+
+In this design section, we focused more on the launcher resiliency under
+`KubeflowDAGRunner`. In `AirflowDAGRunner`, the launcher code is running in the
+same process of Airflow orchestrator which we rely on Airflow to ensure its
```

Comment:
```
same process of Airflow orchestrator which we rely on Airflow to ensure its
```

in which? I'm not sure what this sentence means: there is a word (or two) missing.

Reply:
Reword the sentence to make it more clear.